gammapy/analysis/config.py

```diff
@@ -0,0 +1,189 @@
+from astropy.coordinates import Angle
+from astropy.time import Time
+from astropy.units import Quantity
+from gammapy.utils.scripts import make_path, read_yaml
+from pydantic import BaseModel, FilePath, validator
+from pydantic.utils import deep_update
+from pathlib import Path
+from typing import List
+from enum import Enum
+import yaml
+
+__all__ = ["Config", "General"]
+
+
+class AngleType(Angle):
+    @classmethod
+    def __get_validators__(cls):
+        yield cls.validate
+
+    @classmethod
+    def validate(cls, v):
+        return AngleType(v)
+
+
+class EnergyType(Quantity):
+    @classmethod
+    def __get_validators__(cls):
+        yield cls.validate
+
+    @classmethod
+    def validate(cls, v):
+        assert isinstance(Quantity(v).to("erg"), Quantity)
+        return EnergyType(v)
+
+
+class TimeType(Time):
+    @classmethod
+    def __get_validators__(cls):
+        yield cls.validate
+
+    @classmethod
+    def validate(cls, v):
+        return TimeType(v)
+
+
+class FrameEnum(str, Enum):
+    icrs = 'icrs'
+    galactic = 'galactic'
+
+
+class BackgroundMethodEnum(str, Enum):
+    reflected = 'reflected'
+
+
+class GammapyBaseModel(BaseModel):
+    class Config:
+        validate_assignment = True
+        extra = 'forbid'
+        json_encoders = {
+            AngleType: lambda v: f"{v.value} {v.unit}",
+            EnergyType: lambda v: f"{v.value} {v.unit}",
+            TimeType: lambda v: f"{v.value}",
+        }
+
+    @classmethod
+    def from_yaml(cls, filename):
+        config = read_yaml(filename)
+        return Config(**config)
+
+    def to_yaml(self):
+        return yaml.dump(yaml.safe_load(self.json()))
+
+    def update_from_dict(self, other):
+        data = deep_update(self.dict(exclude_defaults=True), other.dict(exclude_defaults=True))
+        return Config(**data)
+
+
+class Skydir(GammapyBaseModel):
+    frame: FrameEnum = FrameEnum.icrs
+    lon: AngleType = AngleType("83.633 deg")
+    lat: AngleType = AngleType("22.014 deg")
+
+
+class EnergyAxis(GammapyBaseModel):
+    min: EnergyType = EnergyType("0.1 TeV")
+    max: EnergyType = EnergyType("10 TeV")
+    nbins: int = 30
+
+
+class SpatialCircleRange(GammapyBaseModel):
+    frame: FrameEnum = FrameEnum.icrs
+    lon: AngleType = AngleType("83.633 deg")
```

Comment:
Same here and everywhere. Remove the `AngleType` and other manual type casts, and let the framework do the type cast on assignment.

Maybe use a `SkyCoord` here instead of separate lon / lat / frame?

Reply:
> Maybe use a SkyCoord here instead of separate lon / lat / frame?


I do not see how to use `SkyCoord` if we want to keep the syntax below.
```
obs_cone: {frame: galactic, lon: 83.633 deg, lat: 22.014 deg, radius: 3 deg}
```

Reply:
How about this?
```
class SpatialCircleRange(GammapyBaseModel):
    center: SkyCoordType = None
    radius: AngleType = None
```
which would match https://astropy-regions.readthedocs.io/en/latest/api/regions.CircleSkyRegion.html

Reply:
How about this?
```
class SpatialCircleRange(GammapyBaseModel):
    center: SkyCoordType = None
    radius: AngleType = None
```
which would match https://astropy-regions.readthedocs.io/en/latest/api/regions.CircleSkyRegion.html